src/server/protocol_client.cc

```diff
@@ -0,0 +1,332 @@
+    // run we must not create a new socket. sock_mu_ syncs between the two
+    // functions.
+    if (!cntx->IsCancelled()) {
+      sock_.reset(mythread->CreateSocket());
+      serializer_.reset(new ReqSerializer(sock_.get()));
+    } else {
+      return cntx->GetError();
+    }
+  }
+
+  // We set this timeout because this call blocks other REPLICAOF commands. We don't need it for the
+  // rest of the sync.
+  {
+    uint32_t timeout = sock_->timeout();
+    sock_->set_timeout(connect_timeout_ms.count());
+    LOG(WARNING) << server_context_.Description();
+    RETURN_ON_ERR(sock_->Connect(server_context_.endpoint));
+    sock_->set_timeout(timeout);
+  }
+
+  /* These may help but require additional field testing to learn.
+   int yes = 1;
+   CHECK_EQ(0, setsockopt(sock_->native_handle(), IPPROTO_TCP, TCP_NODELAY, &yes, sizeof(yes)));
+   CHECK_EQ(0, setsockopt(sock_->native_handle(), SOL_SOCKET, SO_KEEPALIVE, &yes, sizeof(yes)));
+
+   int intv = 15;
+   CHECK_EQ(0, setsockopt(sock_->native_handle(), IPPROTO_TCP, TCP_KEEPIDLE, &intv, sizeof(intv)));
+
+   intv /= 3;
+   CHECK_EQ(0, setsockopt(sock_->native_handle(), IPPROTO_TCP, TCP_KEEPINTVL, &intv, sizeof(intv)));
+
+   intv = 3;
+   CHECK_EQ(0, setsockopt(sock_->native_handle(), IPPROTO_TCP, TCP_KEEPCNT, &intv, sizeof(intv)));
+  */
+  auto masterauth = absl::GetFlag(FLAGS_masterauth);
+  if (!masterauth.empty()) {
+    ResetParser(false);
+    RETURN_ON_ERR(SendCommandAndReadResponse(StrCat("AUTH ", masterauth)));
+    RETURN_ON_BAD_RESPONSE(CheckRespIsSimpleReply("OK"));
+  }
+  return error_code{};
+}
+
+void ProtocolClient::CloseSocket() {
+  unique_lock lk(sock_mu_);
+  if (sock_) {
+    sock_->proactor()->Await([this] {
+      if (sock_->IsOpen()) {
+        auto ec = sock_->Shutdown(SHUT_RDWR);
+        LOG_IF(ERROR, ec) << "Could not shutdown socket " << ec;
+      }
+    });
+  }
+}
+
+void ProtocolClient::DefaultErrorHandler(const GenericError& err) {
+  CloseSocket();
+}
+
+io::Result<size_t> ProtocolClient::ReadRespReply(base::IoBuf* buffer, bool copy_msg) {
+  DCHECK(parser_);
+
+  error_code ec;
+  if (!buffer) {
+    buffer = &resp_buf_;
+    buffer->Clear();
+  }
+  last_resp_ = "";
+
+  size_t processed_bytes = 0;
+
+  RedisParser::Result result = RedisParser::OK;
+  while (!ec) {
+    uint32_t consumed;
+    if (buffer->InputLen() == 0 || result == RedisParser::INPUT_PENDING) {
+      io::MutableBytes buf = buffer->AppendBuffer();
+      io::Result<size_t> size_res = sock_->Recv(buf);
+      if (!size_res) {
+        LOG(ERROR) << "Socket error " << size_res.error();
+        return nonstd::make_unexpected(size_res.error());
+      }
+
+      VLOG(2) << "Read master response of " << *size_res << " bytes";
+
+      TouchIoTime();
+      buffer->CommitWrite(*size_res);
+    }
+
+    result = parser_->Parse(buffer->InputBuffer(), &consumed, &resp_args_);
+    if (copy_msg)
+      last_resp_ += std::string_view((char*)buffer->InputBuffer().data(), consumed);
```

Comment:
Apparently you need `reinterpret_cast` in this case, lol. It's fine since I cast into `char*` and we don't reallyy care about encodings here, but it's nicer to be as scary about this as possible I guess.